src/main/java/org/apache/commons/collections4/bloomfilter/ArrayCountingBloomFilter.java

```diff
@@ -0,0 +1,396 @@
+     * The iterator returns indexes in their natural order.
+     *
+     * @return an iterator over the enabled indexes
+     */
+    private PrimitiveIterator.OfInt iterator() {
+        return new IndexIterator();
+    }
+
+    @Override
+    public void merge(final BloomFilter other) {
+        applyAsBloomFilter(other, this::increment);
+    }
+
+    @Override
+    public void merge(final Hasher hasher) {
+        applyAsHasher(hasher, this::increment);
+    }
+
+    @Override
+    public boolean remove(BloomFilter other) {
+        applyAsBloomFilter(other, this::decrement);
+        return isValid();
+    }
+
+    @Override
+    public boolean remove(Hasher hasher) {
+        applyAsHasher(hasher, this::decrement);
+        return isValid();
+    }
+
+    @Override
+    public boolean add(CountingBloomFilter other) {
+        applyAsCountingBloomFilter(other, this::add);
+        return isValid();
+    }
+
+    @Override
+    public boolean subtract(CountingBloomFilter other) {
+        applyAsCountingBloomFilter(other, this::subtract);
+        return isValid();
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * <p><em>Implementation note</em>
+     *
+     * <p>The state transition to invalid is permanent.
+     *
+     * <p>This implementation does not correct negative counts to zero or integer
+     * overflow counts to {@link Integer#MAX_VALUE}. Thus the operation that
+     * generated invalid counts can be reversed by using the complement of the
+     * original operation with the same Bloom filter. This will restore the counts
+     * to the state prior to the invalid operation. Counts can then be extracted
+     * using {@link #forEachCount(BitCountConsumer)}.
+     */
+    @Override
+    public boolean isValid() {
+        return state >= 0;
+    }
+
+    @Override
+    public void forEachCount(BitCountConsumer action) {
+        for (int i = 0; i < counts.length; i++) {
+            if (counts[i] != 0) {
+                action.accept(i, counts[i]);
+            }
+        }
+    }
+
+    /**
+     * Apply the action for each index in the Bloom filter.
+     */
+    private void applyAsBloomFilter(final BloomFilter other, final IntConsumer action) {
+        verifyShape(other);
+        if (other instanceof ArrayCountingBloomFilter) {
+            // Only use the presence of non-zero and not the counts
+            final int[] counts2 = ((ArrayCountingBloomFilter) other).counts;
+            for (int i = 0; i < counts2.length; i++) {
+                if (counts2[i] != 0) {
+                    action.accept(i);
+                }
+            }
+        } else {
+            BitSet.valueOf(other.getBits()).stream().forEach(action);
+        }
+    }
+
+    /**
+     * Apply the action for each index in the hasher.
+     */
+    private void applyAsHasher(final Hasher hasher, final IntConsumer action) {
+        verifyHasher(hasher);
+        toSet(hasher).forEach(i -> action.accept(i));
```

Comment:
It would work. Note you would need a +1 on the size of the long[] and you have to write the long bit back to the array:

```java
// Ensure array is big enough
filter = new long[BloomFilterIndexer.getLongIndex(shape.getNumberOfBits()) + 1];

// ...

// Write back the bit index
if ((target & mask) == 0) {
    filter[BloomFilterIndexer.getLongIndex(arg0)] |= mask;
    child.accept(arg0);
}
```

But you are allocating an array of length 1/64 of the number of bits in the filter. We expect a hasher to provide only k-indexes with k being small (e.g. <30).

Certainly there is a potential optimisation to be had by avoiding the use of a Set with boxed integers.

It would be more space efficient to just use an int[] to store up to k-indices. You check each new index using a for loop through those already observed for complexity O(n^2) as search complexity for each index is O(n). In comparison the search complexity of a TreeSet is O(log n) for O(n log n) overall. But with the boxing of the integers and the construction of the tree it would require some performance analysis on typical values of small k to determine which is best.

For now I can mark it as a TODO in the code to investigate using a different data structure to purify the indexes when the expected count of indexes is small/medium/large. I think your class would work for large. Here we require one for small.

Since this may be useful it could be put into a factory to create an appropriate filter for different sizes:
```java
IntConsumer action = ...
int k = ...
// UniqueFilter would ensure the action is called with unique values for
// an expected total of k invocations.
IntConsumer filteredAction = UniqueFilter.forSize(k, action);
```

Reply:
It would work but this is for a case when you are merging possibly duplicate indices for a Bloom filter with a known shape and filled to the expected number of items in the shape. So you are estimating for a different use case where the Hasher is representing a Bloom filter at capacity.

For the use case here it is for the merge of a single item that has been hashed into the filter. So in this case IIUC the Hasher should return only `shape.getNumberOfHashFunctions()` in the iterator.

However there is no mandate to actually do this and the Hasher could return a lot more indexes.

If we require the filtering of indexes which may be duplicated in an optimal way from a Hasher it seems that the Hasher interface should have an estimateSize() function.

There are 2 cases:
- size is known exactly
- size is an estimate

This is dealt with in the Spliterator interface using the methods:

```java
long estimateSize();
long getExactSizeIfKnown();
```

So we could work with this idea:

1. We change the Hasher to return a spliterator and use the JDK framework. For example a StaticHasher can report that it is a known size, a DynamicHasher may not know its size.

2. We add estimateSize and getExactSizeIfKnown to the Hasher interface

I think option 1 is less clutter in the API. The interface can then have a default method to create an iterator from the spliterator if you want to process items one at a time and allow fast fail.

Thus a BloomFilter.merge(Hasher) operation can obtain the spliterator, get the size and then create a filter to make the indexes unique with an optimal data structure. Or it can skip having to filter the indexes if the Hasher reports that its Spliterator is `Spliterator.DISTINCT`.

Changing the Hasher API is for another PR. For now I will abstract out the UniqueFilter concept that accepts a Hasher and IntConsumer as arguments and ensures only unique values are passed to the consumer. This can be used when you want to merge with all the unique bit indexes in a Hasher.

Reply:
OK. I've refactored the extraction of unique indexes from a hasher to a new helper class. This can be optimised later when the Hasher interface is updated to support providing the following information:

- size (number of indexes)
- whether or not the indexes are distinct

This could be satisfied by changing `Hasher.getBits(Shape)` to create a spliterator and not an iterator.